### PR TITLE
Explicit loader Option

### PR DESCRIPTION
**Pull request by cecinestpasunepipe**
(DIS-2029/DIS-2028)

### CODE REVIEW

File: dissect/target/loaders/log.py [hunk 1]
```diff
@@ -1,13 +1,23 @@
...
 from dissect.target.loader import Loader
 
+logloader_option_hint = "hint"
+
 
 class LogLoader(Loader):
+    """Load separate log files without a target. Usage: target-query /evtx/* -L log -f evtx (-L log -h for options)"""
+
+    __loader_args__ = {
+        logloader_option_hint: {"help": "Hint for file type"},
```
Comment (Schamper) [inline dissect/target/loaders/log.py:18]: It's not great to have to define these as dictionaries. Can't you re-use some of the code behind `@arg`?

File: dissect/target/loaders/log.py [hunk 2]
```diff
@@ -26,12 +36,15 @@ def detect(path: Path) -> bool:
...
-            ext = self.options.get("hint", entry.suffix.lower()).strip(".")
+            ext = self.options.get(logloader_option_hint, entry.suffix.lower()).strip(".")
             if (mapping := self.LOGS_DIRS.get(ext, None)) is None:
                 continue
-            mapping = str(vfs.path(mapping).joinpath(entry.name))
-            vfs.map_file(mapping, str(entry))
+            mapping = str(mnt.path(mapping).joinpath(entry.name))
+            mnt.map_file(mapping, str(entry))
         target.filesystems.add(vfs)
         target.fs = vfs
```
Comment (Schamper) [inline dissect/target/loaders/log.py:50]: Isn't this the bug that causes #474? :smile:

File: dissect/target/tools/utils.py [hunk 3]
```diff
@@ -261,3 +263,35 @@ def wrapper(*args, **kwargs):
             raise
 
     return wrapper
+
+
+def args2uri(targets: list[str], loader_name: str, rest: list[str]) -> list[str]:
```
Comment (Schamper) [inline dissect/target/tools/utils.py:268]:
```suggestion
def args_to_uri(targets: list[str], loader_name: str, rest: list[str]) -> list[str]:
```

File: dissect/target/tools/utils.py [hunk 4]
```diff
@@ -261,3 +263,35 @@ def wrapper(*args, **kwargs):
...
+    """Converts argument-style -L to URI-style
```
Comment (Schamper) [inline dissect/target/tools/utils.py:269]:
```suggestion
    """Converts argument-style ``-L`` to URI-style.
```

File: dissect/target/tools/utils.py [hunk 5]
```diff
@@ -261,3 +263,35 @@ def wrapper(*args, **kwargs):
...
+
+    Turns:
+        target-query /evtxs/* -L log --log-hint="evtx" -f evtx
+    into:
+        target-query "log:///evtxs/?hint=evtx" -f evtx
+
+    For classes providing a __loader_args__ dict with
+    argparse values.
```
Comment (Schamper) [inline dissect/target/tools/utils.py:277]:
```suggestion
    For classes providing a __loader_args__ dict with argparse values.
```

File: dissect/target/tools/utils.py [hunk 6]
```diff
@@ -261,3 +263,35 @@ def wrapper(*args, **kwargs):
             raise
 
     return wrapper
+
+
+def args2uri(targets: list[str], loader_name: str, rest: list[str]) -> list[str]:
+    """Converts argument-style -L to URI-style
+
+    Turns:
```
Comment (Schamper) [inline dissect/target/tools/utils.py:271]: Does this show properly in Sphinx?

File: tests/tools/test_utils.py [hunk 7]
```diff
@@ -28,3 +31,23 @@ def test_persist_execution_report():
...
+    [
+        (["/path/to/somewhere"], "loader", ["--loader-option", "1"], ["loader:///path/to/somewhere?option=1"]),
+        (["/path/to/somewhere"], "loader", ["--loader-option", "2"], ["loader:///path/to/somewhere?option=2"]),
+        (["/path/to/somewhere"], "unknown", ["--unknown-option", "3"], ["unknown:///path/to/somewhere"]),
+        (["/path/to/somewhere"], "loader", ["--ignored-option", "4"], ["loader:///path/to/somewhere"]),
+        (["/path/to/somewhere"], "loader", [], ["loader:///path/to/somewhere"]),
+        (["/path/to/somewhere"], "invalid", [], ["invalid:///path/to/somewhere"]),
+    ],
+)
+def test_args2uri(targets, loader_name, rest, uris):
```
Comment (Poeloe) [inline tests/tools/test_utils.py:47]:
```suggestion
def test_args_to_uri(targets, loader_name, rest, uris):
```

File: dissect/target/tools/utils.py (hunk 6 above)
Comment (Schamper) [reply dissect/target/tools/utils.py:271]: I keep being confused by what the proper Sphinx/rst syntax is... Sometimes you need `::` and sometimes a single `:` but an indentation also works... Anyway thanks for verifying!

File: dissect/target/loaders/log.py [hunk 8]
```diff
@@ -26,12 +34,15 @@ def detect(path: Path) -> bool:
 
     def map(self, target: Target) -> None:
         self.target = target
-        vfs = VirtualFilesystem(case_sensitive=False, alt_separator=target.fs.alt_separator)
+        vfs = RootFilesystem(self.target)
```
Comment (Schamper) [inline dissect/target/loaders/log.py:37]: `target.fs` should already exist here, why create a new `RootFilesystem`?

File: dissect/target/loaders/log.py [hunk 9]
```diff
@@ -1,13 +1,21 @@
...
+from dissect.target.filesystem import RootFilesystem, VirtualFilesystem
 from dissect.target.loader import Loader
+from dissect.target.plugin import arg
+
+logloader_option_hint = "hint"
 
 
+@arg("--log-hint", dest=logloader_option_hint, help="Hint for file type")
 class LogLoader(Loader):
+    """Load separate log files without a target. Usage: target-query /evtx/* -L log -f evtx (-L log -h for options)"""
```
Comment (Schamper) [inline dissect/target/loaders/log.py:17]:
```suggestion
    """Load separate log files without a target. Usage: target-query /evtx/* -L log -f evtx (-L log -h for options)."""
```
Having this long line with the usage in it looks a bit weird, but I guess you do that for the `-l` output?

File: dissect/target/loaders/log.py [hunk 10]
```diff
@@ -1,13 +1,21 @@
...
 from dissect.target import Target
-from dissect.target.filesystem import VirtualFilesystem
+from dissect.target.filesystem import RootFilesystem, VirtualFilesystem
 from dissect.target.loader import Loader
+from dissect.target.plugin import arg
+
+logloader_option_hint = "hint"
 
 
+@arg("--log-hint", dest=logloader_option_hint, help="Hint for file type")
```
Comment (Schamper) [inline dissect/target/loaders/log.py:15]:
```suggestion
@arg("--log-hint", dest=logloader_option_hint, help="hint for file type")
```

File: dissect/target/loaders/remote.py [hunk 11]
```diff
@@ -203,6 +214,14 @@ def info(self) -> list[RemoteStream]:
...
 
 
+@arg(f"--remote-{option_client_key}", dest=option_client_key, help="Private key")
+@arg(f"--remote-{option_client_crt}", dest=option_client_crt, help="Client certificate")
+@arg(f"--remote-{option_server_ca}", dest=option_server_ca, help="Certificate Authority")
+@arg(f"--remote-{option_noverify}", dest=option_noverify, help="No certificate verification")
+@arg(f"--remote-{option_reconnects}", dest=option_reconnects, help="Max number of reconnects")
+@arg(f"--remote-{option_shortreads}", dest=option_shortreads, help="Max limit shortreads")
+@arg(f"--remote-{option_reconnectwait}", dest=option_reconnectwait, help="Max time before reconnection attempt")
+@arg(f"--remote-{option_sockettimeout}", dest=option_sockettimeout, help="Sockettimeout")
```
Comment (Schamper) [inline dissect/target/loaders/remote.py:224]:
```suggestion
@arg(f"--remote-{option_client_key}", dest=option_client_key, help="private key")
@arg(f"--remote-{option_client_crt}", dest=option_client_crt, help="client certificate")
@arg(f"--remote-{option_server_ca}", dest=option_server_ca, help="certificate Authority")
@arg(f"--remote-{option_noverify}", dest=option_noverify, help="no certificate verification")
@arg(f"--remote-{option_reconnects}", dest=option_reconnects, help="max number of reconnects")
@arg(f"--remote-{option_shortreads}", dest=option_shortreads, help="max limit shortreads")
@arg(f"--remote-{option_reconnectwait}", dest=option_reconnectwait, help="max time before reconnection attempt")
@arg(f"--remote-{option_sockettimeout}", dest=option_sockettimeout, help="socket timeout")
```
Argparse help are usually lowercase.

File: dissect/target/tools/utils.py [hunk 12]
```diff
@@ -261,3 +263,24 @@ def wrapper(*args, **kwargs):
...
+
+def args_to_uri(targets: list[str], loader_name: str, rest: list[str]) -> list[str]:
+    """Converts argument-style ``-L`` to URI-style.
+
+    Turns:
+        ``target-query /evtxs/* -L log --log-hint="evtx" -f evtx``
+    into:
+        ``target-query "log:///evtxs/*?hint=evtx" -f evtx``
+
+    For classes providing a __args__ dict with argparse values.
```
Comment (Schamper) [inline dissect/target/tools/utils.py:276]:
```suggestion
    For loaders providing ``@arg()`` arguments.
```

File: dissect/target/loaders/remote.py [hunk 13]
```diff
@@ -15,10 +15,20 @@
 from dissect.target.containers.raw import RawContainer
 from dissect.target.exceptions import LoaderError
 from dissect.target.loader import Loader
+from dissect.target.plugin import arg
 from dissect.target.target import Target
 
 log = logging.getLogger(__name__)
 
+option_client_key = "key"
```
Comment (Schamper) [inline dissect/target/loaders/remote.py:23]: I'm not a big fan of making a constant out of every single string/option. These values won't be used outside of this file, can't we just put the strings in the code? It now requires a lot of scrolling up and down to check the actual values with (IMO) little added benefit of putting them here.

File: dissect/target/tools/utils.py [hunk 14]
```diff
@@ -261,3 +263,24 @@ def wrapper(*args, **kwargs):
...
+    """
+    loader = LOADERS_BY_SCHEME.get(loader_name, None)
```
Comment (Schamper) [inline dissect/target/tools/utils.py:278]: Can you add a unit test for a non-existent loader?

**Review by Schamper — CHANGES_REQUESTED** [review]

File: dissect/target/tools/utils.py (hunk 14 above)
Comment (cecinestpasunepipe) [reply dissect/target/tools/utils.py:278]: Sure...

File: dissect/target/loaders/remote.py (hunk 13 above)
Comment (Schamper) [reply dissect/target/loaders/remote.py:23]: These files are small enough and their uses close enough together that I'm personally not concerned for that.

Comment (cecinestpasunepipe) [reply dissect/target/loaders/remote.py]: ok

**Comment by cecinestpasunepipe** [comment]
Important note: this PR now also covers DIS-2028.

File: dissect/target/loaders/asdf.py [hunk 15]
```diff
@@ -15,6 +15,8 @@
 
 
 class AsdfLoader(Loader):
+    """Load an ASDF target"""
```
Comment (Schamper) [inline dissect/target/loaders/asdf.py:18]:
```suggestion
    """Load an ASDF target."""
```

File: dissect/target/loaders/log.py [hunk 16]
```diff
@@ -1,13 +1,27 @@
...
 import urllib
 from pathlib import Path
 from typing import Union
 
 from dissect.target import Target
 from dissect.target.filesystem import VirtualFilesystem
 from dissect.target.loader import Loader
+from dissect.target.plugin import arg
+
+logloader_option_hint = "hint"
```
Comment (Schamper) [inline dissect/target/loaders/log.py:12]: The comment on remote.py also applies here.

File: dissect/target/loaders/raw.py [hunk 17]
```diff
@@ -6,6 +6,8 @@
 
 
 class RawLoader(Loader):
+    """Load raw binary dumps."""
```
Comment (Schamper) [inline dissect/target/loaders/raw.py:9]:
```suggestion
    """Load raw container files such as disk images."""
```

File: dissect/target/loaders/vmwarevm.py [hunk 18]
```diff
@@ -4,6 +4,8 @@
 
 
 class VmwarevmLoader(VmxLoader):
+    """Load vmwarevm folders"""
```
Comment (Schamper) [inline dissect/target/loaders/vmwarevm.py:7]:
```suggestion
    """Load ``*.vmwarevm`` folders from VMware Fusion."""
```

File: tests/tools/test_utils.py [hunk 19]
```diff
@@ -28,3 +31,23 @@ def test_persist_execution_report():
...
+    [
+        (["/path/to/somewhere"], "loader", ["--loader-option", "1"], ["loader:///path/to/somewhere?option=1"]),
+        (["/path/to/somewhere"], "loader", ["--loader-option", "2"], ["loader:///path/to/somewhere?option=2"]),
+        (["/path/to/somewhere"], "unknown", ["--unknown-option", "3"], ["unknown:///path/to/somewhere"]),
+        (["/path/to/somewhere"], "loader", ["--ignored-option", "4"], ["loader:///path/to/somewhere"]),
+        (["/path/to/somewhere"], "loader", [], ["loader:///path/to/somewhere"]),
+        (["/path/to/somewhere"], "invalid", [], ["invalid:///path/to/somewhere"]),
+    ],
+)
+def test_args_to_uri(targets, loader_name, rest, uris):
```
Comment (Schamper) [inline tests/tools/test_utils.py:47]: Can you add type hints?

**Review by Schamper — CHANGES_REQUESTED** [review]

File: dissect/target/loaders/log.py [hunk 20]
```diff
@@ -26,12 +38,14 @@ def detect(path: Path) -> bool:
 
     def map(self, target: Target) -> None:
         self.target = target
-        vfs = VirtualFilesystem(case_sensitive=False, alt_separator=target.fs.alt_separator)
+        vfs = self.target.fs
+        mnt = VirtualFilesystem(case_sensitive=False, alt_separator=target.fs.alt_separator)
+        target.filesystems.add(mnt)
+        vfs.mount("/", mnt)
```
Comment (Schamper) [inline dissect/target/loaders/log.py:44]:
```suggestion
        mnt = VirtualFilesystem(case_sensitive=False, alt_separator=target.fs.alt_separator)
        target.filesystems.add(mnt)
        target.fs.mount("/", mnt)
```

File: dissect/target/loaders/log.py [hunk 21]
```diff
@@ -26,12 +38,14 @@ def detect(path: Path) -> bool:
 
     def map(self, target: Target) -> None:
         self.target = target
```
Comment (Schamper) [inline dissect/target/loaders/log.py:40]:
```suggestion
```

**Review by Schamper — CHANGES_REQUESTED** [review]

File: dissect/target/loaders/log.py [hunk 22]
```diff
@@ -25,13 +37,12 @@ def detect(path: Path) -> bool:
         return False
 
     def map(self, target: Target) -> None:
-        self.target = target
-        vfs = VirtualFilesystem(case_sensitive=False, alt_separator=target.fs.alt_separator)
+        mnt = VirtualFilesystem(case_sensitive=False, alt_separator=target.fs.alt_separator)
+        target.filesystems.add(mnt)
+        target.fs.mount("/", mnt)
```
Comment (Schamper) [inline dissect/target/loaders/log.py:42]:
```suggestion
        vfs = VirtualFilesystem(case_sensitive=False, alt_separator=target.fs.alt_separator)
        target.filesystems.add(vfs)
        target.fs.mount("/", vfs)
```

File: dissect/target/loaders/log.py [hunk 23]
```diff
@@ -25,13 +37,12 @@ def detect(path: Path) -> bool:
...
         for entry in self.path.parent.glob(self.path.name):
             ext = self.options.get("hint", entry.suffix.lower()).strip(".")
             if (mapping := self.LOGS_DIRS.get(ext, None)) is None:
                 continue
-            mapping = str(vfs.path(mapping).joinpath(entry.name))
-            vfs.map_file(mapping, str(entry))
-        target.filesystems.add(vfs)
-        target.fs = vfs
+            mapping = str(mnt.path(mapping).joinpath(entry.name))
+            mnt.map_file(mapping, str(entry))
```
Comment (Schamper) [inline dissect/target/loaders/log.py:48]:
```suggestion
            mapping = str(vfs(mapping).joinpath(entry.name))
            vfs(mapping, str(entry))
```

**Review by Schamper — CHANGES_REQUESTED** [review]
Only noticed just now that the original variable was called `vfs`, apologies. I see no need to change that.

**Review by Schamper — APPROVED** [review]

**Review by Poeloe — APPROVED** [review]

**Pull request merged**